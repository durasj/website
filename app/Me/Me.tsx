import React from 'react';

import Detail from '../Detail';
import Intro from '../Intro';
import Tile from '../Tile';

import { Projects } from '../Projects';

export default class Me extends React.Component<{}, {opened: string}> {
    constructor(props: {}) {
        super(props);

        this.routing(true);
        setTimeout(() => {
            addEventListener('popstate', this.routing.bind(this));
        });
    }

    public render(): React.ReactElement<{}> {
        const tiles = Projects.map((project, index) => {
            const classes = ['column', (typeof project.size !== 'undefined' ? project.size : 'is-half')];

            return (
                <div key={index} className={classes.join(' ')}>
                    <Tile
                        id={project.id}
                        title={project.title}
                        color={project.color}
                        description={project.description}
                        period={project.period}
                        skills={project.skills}
                        photos={project.photos}
                        animation={project.animation}
                    >
                        {project.content}
                    </Tile>
                </div>
            );
        });

        const bodyContent = (this.state.opened === undefined) ? (
            <div className="hero-body">
                <div className="container">
                    <div className="columns is-multiline is-gapless">
                        {tiles}
                    </div>
                </div>
            </div>
        ) : (
            <Detail project={this.getOpenedProject()}/>
        );

        return (
            <section className={'hero is-fullheight' + (!this.state.opened ? ' is-light ' : ' ') + 'is-bold'}>
                {this.state.opened === undefined ? <Intro /> : ''}
                {bodyContent}

                <div className="hero-foot has-text-centered">
                    Jakub Duras, Web developer, jakub@duras.me
                </div>
            </section>
        );
    }

    private routing(initial = false) {
        const openedId = this.currentlyOpenedFromUrl();

        if (openedId !== undefined) {
            const openedProject = Projects.find((project) => project.id === openedId);

            if (openedProject === undefined) {
                location.href = '/404.html';
            }
        }

        if (initial === true) {
            this.state = {
                opened: openedId,
            };
        } else {
            if (openedId !== this.state.opened) {
                this.setState({ opened: openedId });
            }
        }
    }

    private currentlyOpenedFromUrl() {
        const projectId = location.pathname.substr(1) || location.hash.substr(1);

        return (projectId.length > 0) ? projectId : undefined;
    }

    private open(tileId: string) {
        location.hash = '#' + tileId;
    }

    private getOpenedProject() {
        return Projects.find((project) => project.id === this.state.opened);
    }
}
