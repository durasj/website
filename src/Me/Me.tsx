import React, { Component, ReactElement } from 'react';

import Detail from '../Detail';
import Footer from '../Footer';
import Intro from '../Intro';
import List from '../List';

import { Projects } from '../Projects';

interface IState {
    opened: string,
    openedFromList: boolean,
    olderCommercial: boolean,
    olderNoncommercial: boolean,
}

export default class Me extends Component<{}, IState> {
    private handleRouting: EventListener;
    private handleToggleOlderCommercial;
    private handleToggleOlderNoncommercial;
    private handleOpening;
    private handleClosing;

    constructor(props: {}) {
        super(props);

        this.handleToggleOlderCommercial = this.toggleOlder.bind(this, true);
        this.handleToggleOlderNoncommercial = this.toggleOlder.bind(this, false);
        this.handleRouting = this.routing.bind(this);
        this.handleOpening = this.openProject.bind(this);
        this.handleClosing = this.closeDetail.bind(this);

        this.routing(true);
    }

    public componentWillMount() {
        if (typeof window === 'undefined') {
            return; // SSR support
        }

        window.addEventListener('popstate', this.handleRouting);
    }

    public componentWillUnmount() {
        if (typeof window === 'undefined') {
            return; // SSR support
        }

        window.removeEventListener('popstate', this.handleRouting);
    }

    public shouldComponentUpdate(nextProps, nextState) {
        return nextProps !== this.props || nextState !== this.state;
    }

    public render(): ReactElement<{}> {
        const bodyContent = (this.state.opened === undefined) ? (
            <List
                projects={Projects}
                olderCommercial={this.state.olderCommercial}
                olderNoncommercial={this.state.olderNoncommercial}
                onToggleOlderCommercial={this.handleToggleOlderCommercial}
                onToggleOlderNoncommercial={this.handleToggleOlderNoncommercial}
                onOpen={this.handleOpening}
            />
        ) : (
            <Detail
                project={this.getOpenedProject()}
                onClose={this.handleClosing}
            />
        );

        return (
            <div className='container'>
                {this.state.opened === undefined ? <Intro /> : null}

                {bodyContent}

                <Footer />
            </div>
        );
    }

    private routing(initial = false) {
        let openedId = this.currentlyOpenedFromUrl();

        if (openedId !== undefined) {
            const openedProject = Projects.find((project) => project.id === openedId);

            if (openedProject === undefined) {
                location.assign('/404.html');
                openedId = undefined;
            }
        }

        if (initial === true) {
            this.state = {
                opened: openedId,
                openedFromList: false,
                olderCommercial: false,
                olderNoncommercial: false,
            };
        } else {
            if (openedId !== this.state.opened) {
                this.setState({
                    ...this.state,
                    opened: openedId,
                    openedFromList: !this.state.opened && openedId !== undefined,
                });
            }
        }
    }

    private currentlyOpenedFromUrl() {
        if (typeof window === 'undefined') {
            return; // SSR support
        }

        const projectId = location.pathname.substr(1);

        return (projectId.length > 0) ? projectId : undefined;
    }

    private getOpenedProject() {
        return Projects.find((project) => project.id === this.state.opened);
    }

    private toggleOlder(commercial: boolean) {
        if (commercial) {
            this.setState(
                { ...this.state, olderCommercial: !this.state.olderCommercial },
            );
        } else {
            this.setState(
                { ...this.state, olderNoncommercial: !this.state.olderNoncommercial },
            );
        }
    }

    private openProject(projectId) {
        const openedProject = Projects.find((project) => project.id === projectId);

        history.pushState(
            {},
            openedProject.title,
            '/' + openedProject.id,
        );
        this.routing();
    }

    private closeDetail() {
        if (this.state.openedFromList) {
            history.back();
        } else {
            history.pushState({}, '', '/');
            this.routing();
        }
    }
}
