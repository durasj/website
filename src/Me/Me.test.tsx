import { mount } from 'enzyme';
import IJSDOM from 'jsdom';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import mockProjects from '../../testing/mockProjects';
jest.mock('../content.json', () => mockProjects);
import Me from '../Me';

declare var jsdom: IJSDOM.JSDOM;

it('Renders without project opened', () => {
    window.location.replace = jest.fn();
    const renderer = ShallowRenderer.createRenderer();

    const output = renderer.render(
        <Me />,
    );

    expect(output).toMatchSnapshot();
});

it('Renders with project opened', () => {
    const project = mockProjects[0];
    window.scrollTo = jest.fn();
    const renderer = ShallowRenderer.createRenderer();
    jsdom.reconfigure({ url: 'http://localhost/' + project.id});

    const output = renderer.render(
        <Me />,
    );

    expect(output).toMatchSnapshot();
});

it('Redirects to 404 if project was not found', () => {
    window.location.assign = jest.fn();
    const renderer = ShallowRenderer.createRenderer();
    jsdom.reconfigure({ url: 'http://localhost/404' });

    renderer.render(
        <Me />,
    );

    expect(window.location.assign).toHaveBeenCalled();
});

it('Goes back in history when project is opened and closed from list', () => {
    window.scrollTo = jest.fn();
    history.back = jest.fn();

    const component = mount(
        <Me />,
    );

    jest.useFakeTimers();
    component.find('.list a').first().simulate('click');
    jest.runAllTimers();
    component.update();

    component.find('.detail .header .close').simulate('click');
    jest.runAllTimers();

    expect(history.back).toHaveBeenCalled();
});

it('Pushes new empty state when directly opened detail is closed', () => {
    const project = mockProjects[0];
    window.scrollTo = jest.fn();
    jsdom.reconfigure({ url: 'http://localhost/' + project.id});
    history.pushState = jest.fn();

    const component = mount(
        <Me />,
    );

    jest.useFakeTimers();
    component.find('.detail .header .close').simulate('click');
    jest.runAllTimers();

    expect(history.pushState).toHaveBeenCalledWith({}, '', '/');
});

it('Adds and removes event listener', () => {
    window.scrollTo = jest.fn();
    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    const component = mount(
        <Me />,
    );

    expect(window.addEventListener).toHaveBeenCalledWith(
        'popstate',
        expect.anything(),
    );

    component.unmount();

    expect(window.removeEventListener).toHaveBeenCalledWith(
        'popstate',
        expect.anything(),
    );
});
