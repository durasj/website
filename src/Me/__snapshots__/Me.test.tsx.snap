// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with project opened 1`] = `
<div
  className="container"
>
  <Detail
    onClose={[Function]}
    project={
      Object {
        "archived": false,
        "color": "blue",
        "description": "Something nice",
        "id": "something",
        "link": "https://google.com",
        "linkLabel": "My weekend project",
        "period": "2222",
        "photos": Array [
          Object {
            "caption": "Random photo I found on Google",
            "src": "/google/random.jpg",
          },
          Object {
            "caption": "Funny image from Google",
            "src": "/google/funny.jpg",
          },
        ],
        "skills": Array [
          "all",
          "right",
          "skills",
        ],
        "title": "Something",
        "type": "commercial",
      }
    }
  />
  <Footer />
</div>
`;

exports[`Renders without project opened 1`] = `
<div
  className="container"
>
  <Intro />
  <List
    olderCommercial={false}
    olderNoncommercial={false}
    onOpen={[Function]}
    onToggleOlderCommercial={[Function]}
    onToggleOlderNoncommercial={[Function]}
    projects={
      Array [
        Object {
          "archived": false,
          "color": "blue",
          "description": "Something nice",
          "id": "something",
          "link": "https://google.com",
          "linkLabel": "My weekend project",
          "period": "2222",
          "photos": Array [
            Object {
              "caption": "Random photo I found on Google",
              "src": "/google/random.jpg",
            },
            Object {
              "caption": "Funny image from Google",
              "src": "/google/funny.jpg",
            },
          ],
          "skills": Array [
            "all",
            "right",
            "skills",
          ],
          "title": "Something",
          "type": "commercial",
        },
        Object {
          "archived": true,
          "color": "grey",
          "description": "Something archived",
          "id": "else",
          "period": "2009",
          "photos": Array [
            Object {
              "caption": "Low quality photo",
              "src": "/quality.jpg",
            },
          ],
          "skills": Array [
            "all",
            "wrong",
            "skills",
          ],
          "title": "Else",
          "type": "commercial",
        },
        Object {
          "archived": false,
          "color": "white",
          "description": "Pro bono stuff",
          "id": "charity",
          "period": "2019",
          "photos": Array [
            Object {
              "caption": "Kind people at charity",
              "src": "/kind.jpg",
            },
          ],
          "skills": Array [
            "only",
            "pure",
            "skills",
          ],
          "title": "Charity work",
          "type": "noncommercial",
        },
        Object {
          "archived": true,
          "color": "black",
          "description": "Just Another pro bono stuff",
          "id": "other-charity",
          "period": "2000",
          "skills": Array [
            "some",
            "older",
            "skills",
          ],
          "title": "Other charity",
          "type": "noncommercial",
        },
      ]
    }
  />
  <Footer />
</div>
`;
